<launch>
    <arg name="bag_file_use" default="true" />
    <arg name="bag_file" default="/Monster/dataset/KITTI/odometry/2011_09_30_0018_filter.bag"/>

    <param name="use_sim_time" value="true" />
    <group if="$(arg bag_file_use)">
        <node pkg="rosbag" type="play" name="rosbag" 
            args="--clock -r 0.8 $(arg bag_file)" required="true"/>    
    </group>    

    <include file="$(find rviz_car_model)/launch/default.launch" />

    <arg name="run_mloam" default="true" />
    <arg name="run_mloam_odom" default="true" />
    <arg name="run_mloam_mapping" default="true" />    
    <arg name="run_aloam" default="false" />
    <arg name="run_floam" default="false" />
    <arg name="run_legoloam" default="false" />

    <arg name="data_source" default="bag"/>
    <arg name="result_save" default="false"/>
    <arg name="output_path" default="/home/jjiao/trajectory_results/real_vehicle/kitti/xxxxxxx/"/>
    <arg name="with_ua" default="true" /> <!-- awareness of uncertainty propagation-->    
    <arg name="gf_method" default="gd_float" /> <!-- gd_fix, gd_float, rnd, fps -->
    <arg name="gf_ratio_ini" default="1" />
    <arg name="gnc" default="true" />
    <arg name="debug_mode" default="false" />
    <arg name="loss_mode" default="gmc" /> <!-- huber, gmc-->

    <arg name="config_file" default="$(find mloam)/config/config_realvehicle_kitti.yaml" />
    <group if="$(arg run_mloam)">
        <group if="$(arg run_mloam_odom)">
            <node pkg="mloam" type="mloam_node_rv_kitti" name="mloam_node_rv_kitti" 
                args="-config_file=$(arg config_file)
                    -data_source=$(arg data_source)
                    -result_save=$(arg result_save)
                    -output_path=$(arg output_path)" output="screen">
                <remap from="/base_pose_gt" to="/current_odom" />
            </node>
        </group>
        <group if="$(arg run_mloam_mapping)">
            <node pkg="mloam" type="lidar_mapper_keyframe" name="lidar_mapper_keyframe"
                args="-config_file=$(arg config_file)
                      -result_save=$(arg result_save) 
                      -output_path=$(arg output_path)
                      -with_ua=$(arg with_ua)
                      -gf_method=$(arg gf_method)
                      -gf_ratio_ini=$(arg gf_ratio_ini)
                      -gnc=$(arg gnc)
                      -debug_mode=$(arg debug_mode)
                      -loss_mode=$(arg loss_mode)" output="screen">
                <remap from="/laser_odom" to="/laser_odom_0"/>
            </node>
        </group>
    </group>

    <!-- publish static tf -->
    <!-- x y z yaw pitch roll -->
    <!-- x y z qx qy qz qw -->
    <node name="static_tf_world_base_into_world_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0 1.6 0 0 0 world_base world 10"/> 
    <node name="static_tf_world_base_into_laser" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 world laser 10"/>         
    <node name="static_tf_laser_into_laser" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 laser velo 10"/>        
    <node name="static_tf_base_sensor_into_base_link" pkg="tf" type="static_transform_publisher"
        args="0 0 -1.6 0 0 0 aft_mapped base_link 10"/>                          

    <!-- run aloam  -->
    <group if="$(arg run_aloam)">
        <include file="$(find aloam_velodyne)/launch/aloam_exp.launch">
            <arg name="output_path" default="$(arg output_path)"/>
        </include>       
   </group>        

    <!-- run floam  -->
    <group if="$(arg run_floam)">
        <include file="$(find floam)/launch/floam_exp.launch">
            <arg name="output_path" default="$(arg output_path)"/>
        </include>       
   </group>        

    <!-- run legoloam  -->
    <group if="$(arg run_legoloam)">
        <include file="$(find lego_loam)/launch/legoloam_exp.launch">
            <arg name="output_path" default="$(arg output_path)"/>
        </include>
    </group>  

 </launch>
